File: fix-css-duplicates.js
const fs = require('fs')

function fixCSSDuplicates(inputFile, outputFile = null) {
  if (!outputFile) {
    outputFile = inputFile
  }

  console.log(`Анализ файла: ${inputFile}`)

  // Читаем файл
  const content = fs.readFileSync(inputFile, 'utf8')

  // Простой regex для поиска CSS блоков
  const cssBlockRegex = /([^{}]+)\s*\{([^{}]*)\}/g
  const matches = [...content.matchAll(cssBlockRegex)]

  // Группируем по селекторам
  const selectors = {}

  matches.forEach(match => {
    const selector = match[1].trim()
    const properties = match[2].trim()

    if (!selectors[selector]) {
      selectors[selector] = []
    }
    selectors[selector].push(properties)
  })

  // Находим дублирующиеся селекторы
  const duplicates = Object.entries(selectors).filter(
    ([, props]) => props.length > 1
  )

  console.log(`Найдено дублирующихся селекторов: ${duplicates.length}`)

  if (duplicates.length === 0) {
    console.log('Дублирующиеся селекторы не найдены!')
    return 0
  }

  duplicates.forEach(([selector, props]) => {
    console.log(`  - ${selector} (встречается ${props.length} раз)`)
  })

  // Создаем новый контент, удаляя дублирующиеся селекторы
  let newContent = content

  // Для каждого дублирующегося селектора
  for (const [selector, propertyBlocks] of duplicates) {
    // Объединяем все свойства
    const allProperties = {}

    propertyBlocks.forEach(block => {
      const props = block
        .split(';')
        .map(p => p.trim())
        .filter(p => p)
      props.forEach(prop => {
        if (prop.includes(':')) {
          const [key, ...valueParts] = prop.split(':')
          const value = valueParts.join(':').trim()
          allProperties[key.trim()] = value
        }
      })
    })

    // Создаем новый блок CSS
    const newProperties = Object.entries(allProperties)
      .map(([key, value]) => `  ${key}: ${value};`)
      .join('\n')

    const newBlock = `${selector} {\n${newProperties}\n}`

    // Удаляем все старые вхождения этого селектора
    const escapedSelector = selector.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
    const selectorRegex = new RegExp(`${escapedSelector}\\s*\\{[^{}]*\\}`, 'g')
    newContent = newContent.replace(selectorRegex, '')

    // Добавляем новый блок в конец
    newContent += `\n\n${newBlock}`
  }

  // Убираем лишние пустые строки
  newContent = newContent.replace(/\n\s*\n\s*\n/g, '\n\n')

  // Сохраняем результат
  fs.writeFileSync(outputFile, newContent, 'utf8')

  console.log(`Исправленный файл сохранен: ${outputFile}`)
  console.log(`Обработано дублирующихся селекторов: ${duplicates.length}`)

  return duplicates.length
}

// Проверяем аргументы командной строки
if (process.argv.length < 3) {
  console.log(
    'Использование: node fix-css-duplicates.js <input_file> [output_file]'
  )
  process.exit(1)
}

const inputFile = process.argv[2]
const outputFile = process.argv[3] || null

try {
  const duplicatesFixed = fixCSSDuplicates(inputFile, outputFile)
  console.log(`Успешно исправлено ${duplicatesFixed} дублирующихся селекторов!`)
} catch (error) {
  console.error(`Ошибка: ${error.message}`)
  process.exit(1)
}

File: fix_css_duplicates.py
#!/usr/bin/env python3
"""
CSS Duplicate Selector Fixer

Этот скрипт исправляет дублирующиеся селекторы в CSS файле путем объединения
всех их свойств в один селектор.
"""

import re
import sys
from collections import defaultdict

def parse_css_file(filename):
    """Парсит CSS файл и извлекает селекторы с их свойствами."""
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Удаляем комментарии
    content = re.sub(r'/\*.*?\*/', '', content, flags=re.DOTALL)
    
    # Находим все CSS блоки (селектор + свойства)
    pattern = r'([^{}]+)\s*{\s*([^{}]*)\s*}'
    matches = re.findall(pattern, content)
    
    selectors = defaultdict(list)
    all_blocks = []
    
    for selector, properties in matches:
        selector = selector.strip()
        properties = properties.strip()
        
        if properties:  # Игнорируем пустые блоки
            selectors[selector].append(properties)
            all_blocks.append((selector, properties))
    
    return selectors, all_blocks

def merge_properties(property_blocks):
    """Объединяет несколько блоков свойств в один."""
    all_props = {}
    
    for block in property_blocks:
        # Разбиваем свойства на отдельные объявления
        props = re.split(r';\s*', block)
        for prop in props:
            prop = prop.strip()
            if ':' in prop and prop:
                key, value = prop.split(':', 1)
                key = key.strip()
                value = value.strip()
                all_props[key] = value
    
    # Восстанавливаем CSS формат
    result = []
    for key, value in all_props.items():
        result.append(f'  {key}: {value};')
    
    return '\n'.join(result)

def fix_css_duplicates(input_file, output_file=None):
    """Исправляет дублирующиеся селекторы в CSS файле."""
    if output_file is None:
        output_file = input_file
    
    # Читаем оригинальный файл, сохраняя комментарии и структуру
    with open(input_file, 'r', encoding='utf-8') as f:
        original_content = f.read()
    
    selectors, all_blocks = parse_css_file(input_file)
    
    # Находим дублирующиеся селекторы
    duplicates = {sel: props for sel, props in selectors.items() if len(props) > 1}
    
    print(f"Найдено {len(duplicates)} дублирующихся селекторов:")
    for selector in duplicates.keys():
        print(f"  - {selector}")
    
    # Создаем исправленный контент
    processed_selectors = set()
    new_content = []
    
    # Разбиваем оригинальный контент на части для обработки
    lines = original_content.split('\n')
    i = 0
    
    while i < len(lines):
        line = lines[i].strip()
        
        # Проверяем, начинается ли CSS блок
        if '{' in line and not line.startswith('/*') and not line.startswith('@'):
            # Извлекаем селектор
            selector_line = line.split('{')[0].strip()
            
            if selector_line in duplicates and selector_line not in processed_selectors:
                # Это дублирующийся селектор, который мы еще не обработали
                merged_props = merge_properties(duplicates[selector_line])
                new_content.append(f'{selector_line} {{')
                new_content.append(merged_props)
                new_content.append('}')
                new_content.append('')
                processed_selectors.add(selector_line)
                
                # Пропускаем до конца блока
                brace_count = 1
                i += 1
                while i < len(lines) and brace_count > 0:
                    if '{' in lines[i]:
                        brace_count += lines[i].count('{')
                    if '}' in lines[i]:
                        brace_count -= lines[i].count('}')
                    i += 1
                continue
            elif selector_line in processed_selectors:
                # Пропускаем уже обработанный дублирующийся селектор
                brace_count = 1
                i += 1
                while i < len(lines) and brace_count > 0:
                    if '{' in lines[i]:
                        brace_count += lines[i].count('{')
                    if '}' in lines[i]:
                        brace_count -= lines[i].count('}')
                    i += 1
                continue
        
        # Если это не дублирующийся селектор, добавляем как есть
        new_content.append(lines[i])
        i += 1
    
    # Записываем исправленный файл
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write('\n'.join(new_content))
    
    print(f"Исправленный файл сохранен как: {output_file}")
    return len(duplicates)

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Использование: python fix_css_duplicates.py <input_file> [output_file]")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else None
    
    try:
        duplicates_fixed = fix_css_duplicates(input_file, output_file)
        print(f"Успешно исправлено {duplicates_fixed} дублирующихся селекторов!")
    except Exception as e:
        print(f"Ошибка: {e}")
        sys.exit(1)